propagators/opentelemetry-propagator-aws-xray/src/AWSXRayPropagator.ts

```diff
@@ -0,0 +1,194 @@
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Context,
+  TextMapPropagator,
+  SpanContext,
+  TraceFlags,
+  TextMapSetter,
+  TextMapGetter,
+  isSpanContextValid,
+  isValidSpanId,
+  isValidTraceId,
+  INVALID_TRACEID,
+  INVALID_SPANID,
+  getSpanContext,
+  setSpanContext,
+  INVALID_SPAN_CONTEXT,
+} from '@opentelemetry/api';
+
+export const AWSXRAY_TRACE_ID_HEADER = 'X-Amzn-Trace-Id';
+
+const TRACE_HEADER_DELIMITER = ';';
+const KV_DELIMITER = '=';
+
+const TRACE_ID_KEY = 'Root';
+const TRACE_ID_LENGTH = 35;
+const TRACE_ID_VERSION = '1';
+const TRACE_ID_DELIMITER = '-';
+const TRACE_ID_DELIMITER_INDEX_1 = 1;
+const TRACE_ID_DELIMITER_INDEX_2 = 10;
+const TRACE_ID_FIRST_PART_LENGTH = 8;
+
+const PARENT_ID_KEY = 'Parent';
+
+const SAMPLED_FLAG_KEY = 'Sampled';
+const IS_SAMPLED = '1';
+const NOT_SAMPLED = '0';
+
+/**
+ * Implementation of the AWS X-Ray Trace Header propagation protocol. See <a href=
+ * https://https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader>AWS
+ * Tracing header spec</a>
+ *
+ * An example AWS Xray Tracing Header is shown below:
+ * X-Amzn-Trace-Id: Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1
+ */
+export class AWSXRayPropagator implements TextMapPropagator {
+  inject(context: Context, carrier: unknown, setter: TextMapSetter) {
+    const spanContext = getSpanContext(context);
+    if (!spanContext || !isSpanContextValid(spanContext)) return;
+
+    const otTraceId = spanContext.traceId;
+    const timestamp = otTraceId.substring(0, TRACE_ID_FIRST_PART_LENGTH);
+    const randomNumber = otTraceId.substring(TRACE_ID_FIRST_PART_LENGTH);
+
+    const parentId = spanContext.spanId;
+    const samplingFlag =
+      (TraceFlags.SAMPLED & spanContext.traceFlags) === TraceFlags.SAMPLED
+        ? IS_SAMPLED
+        : NOT_SAMPLED;
+    // TODO: Add OT trace state to the X-Ray trace header
+
+    const traceHeader = `Root=1-${timestamp}-${randomNumber};Parent=${parentId};Sampled=${samplingFlag}`;
+    setter.set(carrier, AWSXRAY_TRACE_ID_HEADER, traceHeader);
+  }
+
+  extract(context: Context, carrier: unknown, getter: TextMapGetter): Context {
+    const spanContext = this.getSpanContextFromHeader(carrier, getter);
+    if (!isSpanContextValid(spanContext)) return context;
+
+    return setSpanContext(context, spanContext);
+  }
+
+  fields(): string[] {
+    return [AWSXRAY_TRACE_ID_HEADER];
+  }
+
+  private getSpanContextFromHeader(
+    carrier: unknown,
+    getter: TextMapGetter
+  ): SpanContext {
+    const traceHeader = getter.get(carrier, AWSXRAY_TRACE_ID_HEADER);
```

Comment:
I think this wont work for NodeJs HTTP headers because all incoming header names are converted to lowercase but `AWSXRAY_TRACE_ID_HEADER` is mixed case.
Or is it intended that users use a custom getter here? If yes this should be mentioned in docs.

Reply:
Hmm - I figured issues about lowercase vs uppercase would be handled at a different layer than the propagator. I assume that means it's ok for me to use all lowercase in this propagator, similar to the b3 one

https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-propagator-b3/src/constants.ts#L21

So switched to it